File: xim/poc/SkeletonAnimator.kt
package xim.poc

import xim.math.Quaternion
import xim.math.Vector3f
import xim.resource.DatId
import xim.resource.SkeletonAnimation
import xim.resource.SkeletonAnimationKeyFrameTransform
import xim.resource.SkeletonAnimationResource
import kotlin.math.max

class AnimationSnapshot {

    private val jointSnapshots = HashMap<Int, SkeletonAnimationKeyFrameTransform>()

    constructor(previous: SkeletonAnimationContext) {
        val joints = previous.animation.keyFrameSets.keys
        for (joint in joints) {
            val snapshot = previous.animation.getJointTransform(joint, previous.currentFrame) ?: continue
            jointSnapshots[joint] = snapshot
        }
    }

    constructor(transition: AnimationTransition) {
        val allJoints = transition.previous.jointSnapshots.keys + transition.next.animation.keyFrameSets.keys
        for (joint in allJoints) {
            val snapshot = transition.getJointTransform(joint) ?: continue
            jointSnapshots[joint] = snapshot
        }
    }

    fun getJointTransform(jointIndex: Int): SkeletonAnimationKeyFrameTransform? {
        return jointSnapshots[jointIndex]
    }

}

class AnimationTransition(val previous: AnimationSnapshot, val next: SkeletonAnimationContext, val transitionDuration: Float, val inBetween: SkeletonAnimationResource?) {

    companion object {
        private val unitTransform = SkeletonAnimationKeyFrameTransform()
    }

    private var progress = 0f

    fun update(elapsedFrames: Float): Boolean {
        progress += elapsedFrames
        return isComplete()
    }

    fun isComplete(): Boolean {
        return progress >= transitionDuration
    }

    fun getJointTransform(jointIndex: Int): SkeletonAnimationKeyFrameTransform? {
        val t = progress/transitionDuration

        return if (inBetween == null) {
            val keyFramePrev = previous.getJointTransform(jointIndex)
            val keyFrameNext = next.getJointTransform(jointIndex)
            interpolate(keyFramePrev, keyFrameNext, t)
        } else if (t < 0.5f) {
            val keyFramePrev = previous.getJointTransform(jointIndex)
            val keyFrameNext = inBetween.skeletonAnimation.getJointTransform(jointIndex, 0f)
            interpolate(keyFramePrev, keyFrameNext, t * 2f)
        } else {
            val keyFramePrev = inBetween.skeletonAnimation.getJointTransform(jointIndex, 0f)
            val keyFrameNext = next.getJointTransform(jointIndex)
            interpolate(keyFramePrev, keyFrameNext, (t - 0.5f) * 2f)
        }
    }

    private fun interpolate(a: SkeletonAnimationKeyFrameTransform?, b: SkeletonAnimationKeyFrameTransform?, delta: Float) : SkeletonAnimationKeyFrameTransform? {
        if (a == null && b == null) { return null }
        if (a != null && b != null) { return SkeletonAnimationKeyFrameTransform.interpolate(a, b, delta) }

        // Interpolating from the "unit" transform tends to be smoother when the previous animation doesn't act on the joint
        // However, doing this for `scale` breaks some effects (ranged-attack while engaged), so just default to whichever isn't null
        val rotation = Quaternion.nlerp(a?.rotation ?: unitTransform.rotation, b?.rotation ?: unitTransform.rotation, delta)
        val translation = Vector3f.lerp(a?.translation ?: unitTransform.translation, b?.translation ?: unitTransform.translation, delta)
        val scale = a?.scale ?: b?.scale ?: unitTransform.scale

        return SkeletonAnimationKeyFrameTransform(rotation = rotation, translation = translation, scale = scale)
    }

}

data class LoopParams (
    val loopDuration: Float?,
    val numLoops: Int?,
    val lowPriority: Boolean = false,
) {
    companion object {
        fun lowPriorityLoop(): LoopParams = LoopParams(loopDuration = null, numLoops = null, lowPriority = true)
    }
}

class TransitionParams (
    val transitionInTime: Float = 7.5f,
    val transitionOutTime: Float = 7.5f,
    val inBetween: DatId? = null,
    var eagerTransitionOut: Boolean = false,
) {
    var resolvedInBetween: Map<Int, SkeletonAnimationResource>? = null
}

class SkeletonAnimationContext(
    val animation: SkeletonAnimation,
    val loopParams: LoopParams,
    val transitionParams: TransitionParams?,
    val modelSlotVisibilityState: ModelSlotVisibilityState?,
) {

    var currentFrame = 0f
    var framesSinceComplete = 0f
    var totalLifeTime = 0f

    private var completed = false
    private var loopCounter = 0

    fun advance(elapsedFrames: Float) {
        totalLifeTime += elapsedFrames
        if (completed || transitionParams?.eagerTransitionOut == true) {
            framesSinceComplete += elapsedFrames
        }

        if (loopParams.loopDuration == 0f) {
            currentFrame = 0f
            completed = true
            return
        }

        val loopDuration = loopParams.loopDuration ?: animation.getLengthInFrames()
        val scalingFactor = animation.getLengthInFrames() / loopDuration

        currentFrame += (elapsedFrames * scalingFactor)
        currentFrame = applyLoopBounds()
    }

    fun getJointTransform(jointIndex: Int): SkeletonAnimationKeyFrameTransform? {
        return animation.getJointTransform(jointIndex, currentFrame)
    }

    fun isDoneLooping(): Boolean {
        return loopParams.numLoops == null || completed
    }

    private fun applyLoopBounds(): Float {
        val maxLoops = loopParams.numLoops ?: 0

        while (currentFrame > animation.getLengthInFrames()) {
            loopCounter += 1
            currentFrame -= animation.getLengthInFrames()
        }

        if (maxLoops != 0 && loopCounter >= maxLoops) {
            completed = true
            return animation.getLengthInFrames()
        }

        return currentFrame
    }

}

class SkeletonAnimator(private val animationSlot: Int) {

    var currentAnimation: SkeletonAnimationContext? = null
    var transition: AnimationTransition? = null

    fun update(elapsedFrames: Float) {
        if (transition?.update(elapsedFrames) == true) {
            transition = null
        } else {
            currentAnimation?.advance(elapsedFrames)
        }
    }

    fun setNextAnimation(skeletonAnimationContext: SkeletonAnimationContext, transitionParams: TransitionParams?) {
        val current = currentAnimation

        if (current == null || transitionParams?.transitionInTime == 0f) {
            currentAnimation = skeletonAnimationContext
            return
        }

        if ((current.animation === skeletonAnimationContext.animation) && current.loopParams.lowPriority) {
            return
        }

        // TODO - disabling same-slot-interpolation for Slot[5] helps with various "triplet" mobs (Cluster, Bats, etc)
        // Generally, animations in Slot[5] want cross-slot-interpolation instead
        if (animationSlot != 5) {
            val transitionDuration = if (transitionParams != null) {
                transitionParams.transitionInTime
            } else if (current.transitionParams != null && current.transitionParams.transitionOutTime > 0f) {
                current.transitionParams.transitionOutTime
            } else {
                7.5f
            }

            val snapshot = transition?.let { AnimationSnapshot(it) } ?: AnimationSnapshot(current)
            val maybeInBetweenFrame = transitionParams?.resolvedInBetween?.get(animationSlot)

            transition = AnimationTransition(snapshot, skeletonAnimationContext, transitionDuration, maybeInBetweenFrame)
        }

        currentAnimation = skeletonAnimationContext
    }

    fun getJointTransform(jointIndex: Int): SkeletonAnimationKeyFrameTransform? {
        return if (transition != null) {
            transition?.getJointTransform(jointIndex)
        } else {
            currentAnimation?.getJointTransform(jointIndex)
        }
    }

}

class SkeletonAnimationCoordinator {

    val animations = Array<SkeletonAnimator?>(8) { null }

    fun update(elapsedFrames: Float) {
        animations.forEach { it?.update(elapsedFrames) }
    }

    fun registerAnimation(skeletonAnimationResources: List<SkeletonAnimationResource>, loopParams: LoopParams, transitionParams: TransitionParams? = null, modelSlotVisibilityState: ModelSlotVisibilityState? = null, overrideCondition: (SkeletonAnimator) -> Boolean = { true }) {
        for (skeletonAnimationResource in skeletonAnimationResources) {
            val animationType = skeletonAnimationResource.id.finalDigit() ?: 0
            val animator = getOrPut(animationType) { SkeletonAnimator(animationType) }
            val context = SkeletonAnimationContext(skeletonAnimationResource.skeletonAnimation, loopParams, transitionParams, modelSlotVisibilityState)

            if (overrideCondition(animator)) {
                animator.setNextAnimation(context, transitionParams)
            }
        }
    }

    fun hasCompleteTransitionOutAnimations(): Boolean {
        return animations.any { it != null && readyForTransitionOut(it, requireTransitionOut = true) }
    }

    fun registerIdleAnimation(skeletonAnimationResources: List<SkeletonAnimationResource>, requireTransitionOut: Boolean) {
        registerAnimation(skeletonAnimationResources, loopParams = LoopParams.lowPriorityLoop()) {
            readyForTransitionOut(it, requireTransitionOut)
        }
    }

    fun clearCompleteAnimations() {
        for (i in animations.indices) {
            val current = animations[i] ?: continue
            val animation = current.currentAnimation ?: continue

            if (animation.loopParams.lowPriority) { continue }
            if (!readyForTransitionOut(current, requireTransitionOut = true)) { continue }

            val transitionOutTime = animation.transitionParams?.transitionOutTime ?: continue
            if (animation.framesSinceComplete >= transitionOutTime) { animations[i] = null }
        }
    }

    fun getSlotVisibilityOverrides(): List<ModelSlotVisibilityState> {
        return animations.mapNotNull { it?.currentAnimation }.mapNotNull { it.modelSlotVisibilityState }
    }

    private fun readyForTransitionOut(animator: SkeletonAnimator, requireTransitionOut: Boolean): Boolean {
        val current = animator.currentAnimation

        val transitionOutReqs = if (!requireTransitionOut) { true } else {
            val outTime = current?.transitionParams?.transitionOutTime
            val noOutTimeCase = (outTime == null || outTime == 0f)
            val outTimeCase = outTime != null && outTime > 0f
            noOutTimeCase || outTimeCase
        }

        var doneLooping = current?.loopParams == null || current.isDoneLooping()

        val eagerOut = current?.transitionParams?.eagerTransitionOut
        if (eagerOut == true) { doneLooping = true }

        return transitionOutReqs && doneLooping
    }

    fun getJointTransform(jointIndex: Int): SkeletonAnimationKeyFrameTransform? {
        var highTransform: SkeletonAnimationKeyFrameTransform? = null
        var highAnimator: SkeletonAnimator? = null
        var lowTransform: SkeletonAnimationKeyFrameTransform? = null

        for (i in 8 downTo 0) {
            val animator = animations[i] ?: continue
            val animatorTransform = animator.getJointTransform(jointIndex) ?: continue

            if (highTransform == null) {
                highTransform = animatorTransform
                highAnimator = animator
            } else if (lowTransform == null) {
                lowTransform = animatorTransform
                break
            }
        }

        return if (highTransform != null && highAnimator != null && lowTransform != null) {
            crossSlotInterpolation(highTransform, highAnimator, lowTransform)
        } else {
            highTransform ?: lowTransform
        }
    }

    private fun crossSlotInterpolation(highSlot: SkeletonAnimationKeyFrameTransform, highAnimator: SkeletonAnimator, lowSlot: SkeletonAnimationKeyFrameTransform): SkeletonAnimationKeyFrameTransform {
        val highAnimation = highAnimator.currentAnimation ?: return highSlot
        val transitionParams = highAnimation.transitionParams ?: return highSlot

        var deltaIn = 0f
        var deltaOut = 0f

        if (highAnimation.totalLifeTime < transitionParams.transitionInTime) {
            deltaIn = (1f - highAnimation.totalLifeTime/transitionParams.transitionInTime)
        }

        if (readyForTransitionOut(highAnimator, requireTransitionOut = true)) {
            deltaOut = if (transitionParams.transitionOutTime == 0f) { 0f } else { highAnimation.framesSinceComplete/transitionParams.transitionOutTime }
        }

        val delta = max(deltaIn, deltaOut)

        return if (delta <= 0f) {
            highSlot
        } else if (delta >= 1f) {
            lowSlot
        } else {
            SkeletonAnimationKeyFrameTransform.interpolate(highSlot, lowSlot, delta)
        }
    }

    fun isTransitioning(): Boolean {
        return animations.any { it?.transition != null }
    }

    fun clear() {
        for (i in animations.indices) { animations[i] = null }
    }

    private fun getOrPut(slot: Int, fn: () -> SkeletonAnimator): SkeletonAnimator {
        val current = animations[slot]
        if (current != null) { return current }

        val new = fn.invoke()
        animations[slot] = new
        return new
    }

}

File: lib/runtime/skeletonAnimator.ts
import {
  DatId,
  type Quaternion,
  type SkeletonAnimation,
  type SkeletonAnimationKeyFrameTransform,
  type SkeletonAnimationResource,
  type Vector3,
} from '~/lib/resource/datResource'
import { datIdFinalDigit } from '~/lib/runtime/datId'
import type { ModelSlotVisibilityState } from '~/lib/runtime/actorModel'

const unitTransform: SkeletonAnimationKeyFrameTransform = {
  rotation: { x: 0, y: 0, z: 0, w: 1 },
  translation: { x: 0, y: 0, z: 0 },
  scale: { x: 1, y: 1, z: 1 },
}

function normalizeQuaternion(q: Quaternion): Quaternion {
  const mag = Math.hypot(q.x, q.y, q.z, q.w)
  if (mag <= 1e-7) {
    return { ...unitTransform.rotation }
  }
  return {
    x: q.x / mag,
    y: q.y / mag,
    z: q.z / mag,
    w: q.w / mag,
  }
}

function nlerpQuaternion(a: Quaternion, b: Quaternion, t: number): Quaternion {
  const dot = a.x * b.x + a.y * b.y + a.z * b.z + a.w * b.w
  const rhs = dot < 0
    ? { x: -b.x, y: -b.y, z: -b.z, w: -b.w }
    : b

  return normalizeQuaternion({
    x: a.x * (1 - t) + rhs.x * t,
    y: a.y * (1 - t) + rhs.y * t,
    z: a.z * (1 - t) + rhs.z * t,
    w: a.w * (1 - t) + rhs.w * t,
  })
}

function lerpVector3(a: Vector3, b: Vector3, t: number): Vector3 {
  return {
    x: a.x + (b.x - a.x) * t,
    y: a.y + (b.y - a.y) * t,
    z: a.z + (b.z - a.z) * t,
  }
}

function interpolateTransform(
  a: SkeletonAnimationKeyFrameTransform | null,
  b: SkeletonAnimationKeyFrameTransform | null,
  delta: number,
): SkeletonAnimationKeyFrameTransform | null {
  if (!a && !b) {
    return null
  }

  if (a && b) {
    return {
      rotation: nlerpQuaternion(a.rotation, b.rotation, delta),
      translation: lerpVector3(a.translation, b.translation, delta),
      scale: lerpVector3(a.scale, b.scale, delta),
    }
  }

  const only = a ?? b
  if (!only) {
    return null
  }

  return {
    rotation: nlerpQuaternion(a?.rotation ?? unitTransform.rotation, b?.rotation ?? unitTransform.rotation, delta),
    translation: lerpVector3(a?.translation ?? unitTransform.translation, b?.translation ?? unitTransform.translation, delta),
    scale: only.scale,
  }
}

export class AnimationSnapshot {
  private readonly jointSnapshots = new Map<number, SkeletonAnimationKeyFrameTransform>()

  constructor(previous: SkeletonAnimationContext)
  constructor(previous: AnimationTransition)
  constructor(previous: SkeletonAnimationContext | AnimationTransition) {
    if (previous instanceof AnimationTransition) {
      const allJoints = new Set<number>()
      for (const joint of previous.previous.jointSnapshots.keys()) {
        allJoints.add(joint)
      }
      for (const joint of previous.next.animation.keyFrameSets.keys()) {
        allJoints.add(joint)
      }

      for (const joint of allJoints) {
        const snapshot = previous.getJointTransform(joint)
        if (snapshot) {
          this.jointSnapshots.set(joint, snapshot)
        }
      }

      return
    }

    for (const joint of previous.animation.keyFrameSets.keys()) {
      const snapshot = previous.animation.getJointTransform(joint, previous.currentFrame)
      if (snapshot) {
        this.jointSnapshots.set(joint, snapshot)
      }
    }
  }

  getJointTransform(jointIndex: number): SkeletonAnimationKeyFrameTransform | null {
    return this.jointSnapshots.get(jointIndex) ?? null
  }
}

export class AnimationTransition {
  readonly previous: AnimationSnapshot
  readonly next: SkeletonAnimationContext
  readonly transitionDuration: number
  readonly inBetween: SkeletonAnimationResource | null

  private progress = 0

  constructor(previous: AnimationSnapshot, next: SkeletonAnimationContext, transitionDuration: number, inBetween: SkeletonAnimationResource | null) {
    this.previous = previous
    this.next = next
    this.transitionDuration = Math.max(0.0001, transitionDuration)
    this.inBetween = inBetween
  }

  update(elapsedFrames: number): boolean {
    this.progress += elapsedFrames
    return this.isComplete()
  }

  isComplete(): boolean {
    return this.progress >= this.transitionDuration
  }

  getJointTransform(jointIndex: number): SkeletonAnimationKeyFrameTransform | null {
    const t = this.progress / this.transitionDuration

    if (!this.inBetween) {
      return interpolateTransform(
        this.previous.getJointTransform(jointIndex),
        this.next.getJointTransform(jointIndex),
        t,
      )
    }

    if (t < 0.5) {
      return interpolateTransform(
        this.previous.getJointTransform(jointIndex),
        this.inBetween.animation.getJointTransform(jointIndex, 0),
        t * 2,
      )
    }

    return interpolateTransform(
      this.inBetween.animation.getJointTransform(jointIndex, 0),
      this.next.getJointTransform(jointIndex),
      (t - 0.5) * 2,
    )
  }
}

export class LoopParams {
  readonly loopDuration: number | null
  readonly numLoops: number | null
  readonly lowPriority: boolean

  constructor(loopDuration: number | null, numLoops: number | null, lowPriority = false) {
    this.loopDuration = loopDuration
    this.numLoops = numLoops
    this.lowPriority = lowPriority
  }

  static lowPriorityLoop(): LoopParams {
    return new LoopParams(null, null, true)
  }
}

export class TransitionParams {
  readonly transitionInTime: number
  readonly transitionOutTime: number
  readonly inBetween: DatId | null
  eagerTransitionOut: boolean
  resolvedInBetween: Map<number, SkeletonAnimationResource> | null = null

  constructor(
    transitionInTime = 7.5,
    transitionOutTime = 7.5,
    inBetween: DatId | null = null,
    eagerTransitionOut = false,
  ) {
    this.transitionInTime = transitionInTime
    this.transitionOutTime = transitionOutTime
    this.inBetween = inBetween
    this.eagerTransitionOut = eagerTransitionOut
  }
}

export class SkeletonAnimationContext {
  readonly animation: SkeletonAnimation
  readonly loopParams: LoopParams
  readonly transitionParams: TransitionParams | null
  readonly modelSlotVisibilityState: ModelSlotVisibilityState | null

  currentFrame = 0
  framesSinceComplete = 0
  totalLifeTime = 0

  private completed = false
  private loopCounter = 0

  constructor(
    animation: SkeletonAnimation,
    loopParams: LoopParams,
    transitionParams: TransitionParams | null,
    modelSlotVisibilityState: ModelSlotVisibilityState | null,
  ) {
    this.animation = animation
    this.loopParams = loopParams
    this.transitionParams = transitionParams
    this.modelSlotVisibilityState = modelSlotVisibilityState
  }

  advance(elapsedFrames: number): void {
    this.totalLifeTime += elapsedFrames
    if (this.completed || this.transitionParams?.eagerTransitionOut === true) {
      this.framesSinceComplete += elapsedFrames
    }

    if (this.loopParams.loopDuration === 0) {
      this.currentFrame = 0
      this.completed = true
      return
    }

    const loopDuration = this.loopParams.loopDuration ?? this.animation.getLengthInFrames()
    const scalingFactor = this.animation.getLengthInFrames() / loopDuration
    this.currentFrame += elapsedFrames * scalingFactor
    this.currentFrame = this.applyLoopBounds()
  }

  getJointTransform(jointIndex: number): SkeletonAnimationKeyFrameTransform | null {
    return this.animation.getJointTransform(jointIndex, this.currentFrame)
  }

  isDoneLooping(): boolean {
    return this.loopParams.numLoops === null || this.completed
  }

  private applyLoopBounds(): number {
    const maxLoops = this.loopParams.numLoops ?? 0
    while (this.currentFrame > this.animation.getLengthInFrames()) {
      this.loopCounter += 1
      this.currentFrame -= this.animation.getLengthInFrames()
    }

    if (maxLoops !== 0 && this.loopCounter >= maxLoops) {
      this.completed = true
      return this.animation.getLengthInFrames()
    }

    return this.currentFrame
  }
}

export class SkeletonAnimator {
  readonly animationSlot: number
  currentAnimation: SkeletonAnimationContext | null = null
  transition: AnimationTransition | null = null

  constructor(animationSlot: number) {
    this.animationSlot = animationSlot
  }

  update(elapsedFrames: number): void {
    if (this.transition?.update(elapsedFrames)) {
      this.transition = null
      return
    }

    this.currentAnimation?.advance(elapsedFrames)
  }

  setNextAnimation(skeletonAnimationContext: SkeletonAnimationContext, transitionParams: TransitionParams | null): void {
    const current = this.currentAnimation
    if (!current || transitionParams?.transitionInTime === 0) {
      this.currentAnimation = skeletonAnimationContext
      return
    }

    if (current.animation === skeletonAnimationContext.animation && current.loopParams.lowPriority) {
      return
    }

    if (this.animationSlot !== 5) {
      const transitionDuration = transitionParams?.transitionInTime
        ?? (current.transitionParams && current.transitionParams.transitionOutTime > 0 ? current.transitionParams.transitionOutTime : 7.5)

      const snapshot = this.transition ? new AnimationSnapshot(this.transition) : new AnimationSnapshot(current)
      const maybeInBetweenFrame = transitionParams?.resolvedInBetween?.get(this.animationSlot) ?? null
      this.transition = new AnimationTransition(snapshot, skeletonAnimationContext, transitionDuration, maybeInBetweenFrame)
    }

    this.currentAnimation = skeletonAnimationContext
  }

  getJointTransform(jointIndex: number): SkeletonAnimationKeyFrameTransform | null {
    if (this.transition) {
      return this.transition.getJointTransform(jointIndex)
    }
    return this.currentAnimation?.getJointTransform(jointIndex) ?? null
  }
}

export class SkeletonAnimationCoordinator {
  readonly animations: Array<SkeletonAnimator | null> = Array.from({ length: 8 }, () => null)

  update(elapsedFrames: number): void {
    for (const animation of this.animations) {
      animation?.update(elapsedFrames)
    }
  }

  registerAnimation(
    skeletonAnimationResources: readonly SkeletonAnimationResource[],
    loopParams: LoopParams,
    transitionParams: TransitionParams | null = null,
    modelSlotVisibilityState: ModelSlotVisibilityState | null = null,
    overrideCondition: (animator: SkeletonAnimator) => boolean = () => true,
  ): void {
    for (const skeletonAnimationResource of skeletonAnimationResources) {
      const animationType = datIdFinalDigit(skeletonAnimationResource.id) ?? 0
      const animator = this.getOrPut(animationType)
      const context = new SkeletonAnimationContext(
        skeletonAnimationResource.animation,
        loopParams,
        transitionParams,
        modelSlotVisibilityState,
      )

      if (overrideCondition(animator)) {
        animator.setNextAnimation(context, transitionParams)
      }
    }
  }

  hasCompleteTransitionOutAnimations(): boolean {
    return this.animations.some((animation) => animation !== null && this.readyForTransitionOut(animation, true))
  }

  registerIdleAnimation(
    skeletonAnimationResources: readonly SkeletonAnimationResource[],
    requireTransitionOut: boolean,
  ): void {
    this.registerAnimation(
      skeletonAnimationResources,
      LoopParams.lowPriorityLoop(),
      null,
      null,
      (animator) => this.readyForTransitionOut(animator, requireTransitionOut),
    )
  }

  clearCompleteAnimations(): void {
    for (let i = 0; i < this.animations.length; i += 1) {
      const current = this.animations[i]
      const animation = current?.currentAnimation
      if (!current || !animation || animation.loopParams.lowPriority) {
        continue
      }
      if (!this.readyForTransitionOut(current, true)) {
        continue
      }

      const transitionOutTime = animation.transitionParams?.transitionOutTime
      if (transitionOutTime !== undefined && animation.framesSinceComplete >= transitionOutTime) {
        this.animations[i] = null
      }
    }
  }

  getSlotVisibilityOverrides(): ModelSlotVisibilityState[] {
    return this.animations
      .map((animation) => animation?.currentAnimation)
      .filter((value): value is SkeletonAnimationContext => value !== null && value !== undefined)
      .map((value) => value.modelSlotVisibilityState)
      .filter((value): value is ModelSlotVisibilityState => value !== null)
  }

  getJointTransform(jointIndex: number): SkeletonAnimationKeyFrameTransform | null {
    let highTransform: SkeletonAnimationKeyFrameTransform | null = null
    let highAnimator: SkeletonAnimator | null = null
    let lowTransform: SkeletonAnimationKeyFrameTransform | null = null

    for (let i = this.animations.length - 1; i >= 0; i -= 1) {
      const animator = this.animations[i]
      if (!animator) {
        continue
      }

      const transform = animator.getJointTransform(jointIndex)
      if (!transform) {
        continue
      }

      if (!highTransform) {
        highTransform = transform
        highAnimator = animator
      } else if (!lowTransform) {
        lowTransform = transform
        break
      }
    }

    if (highTransform && highAnimator && lowTransform) {
      return this.crossSlotInterpolation(highTransform, highAnimator, lowTransform)
    }

    return highTransform ?? lowTransform
  }

  isTransitioning(): boolean {
    return this.animations.some((animation) => animation?.transition !== null)
  }

  clear(): void {
    for (let i = 0; i < this.animations.length; i += 1) {
      this.animations[i] = null
    }
  }

  private crossSlotInterpolation(
    highSlot: SkeletonAnimationKeyFrameTransform,
    highAnimator: SkeletonAnimator,
    lowSlot: SkeletonAnimationKeyFrameTransform,
  ): SkeletonAnimationKeyFrameTransform {
    const highAnimation = highAnimator.currentAnimation
    const transitionParams = highAnimation?.transitionParams
    if (!highAnimation || !transitionParams) {
      return highSlot
    }

    let deltaIn = 0
    let deltaOut = 0

    if (highAnimation.totalLifeTime < transitionParams.transitionInTime) {
      deltaIn = 1 - (highAnimation.totalLifeTime / transitionParams.transitionInTime)
    }

    if (this.readyForTransitionOut(highAnimator, true)) {
      deltaOut = transitionParams.transitionOutTime === 0
        ? 0
        : highAnimation.framesSinceComplete / transitionParams.transitionOutTime
    }

    const delta = Math.max(deltaIn, deltaOut)
    if (delta <= 0) {
      return highSlot
    }
    if (delta >= 1) {
      return lowSlot
    }

    return {
      rotation: nlerpQuaternion(highSlot.rotation, lowSlot.rotation, delta),
      translation: lerpVector3(highSlot.translation, lowSlot.translation, delta),
      scale: lerpVector3(highSlot.scale, lowSlot.scale, delta),
    }
  }

  private readyForTransitionOut(animator: SkeletonAnimator, requireTransitionOut: boolean): boolean {
    const current = animator.currentAnimation
    const transitionOutReqs = !requireTransitionOut || (() => {
      const outTime = current?.transitionParams?.transitionOutTime
      return outTime === undefined || outTime >= 0
    })()

    let doneLooping = current?.loopParams === undefined || current.isDoneLooping()
    if (current?.transitionParams?.eagerTransitionOut) {
      doneLooping = true
    }

    return transitionOutReqs && doneLooping
  }

  private getOrPut(slot: number): SkeletonAnimator {
    const normalizedSlot = Math.max(0, Math.min(this.animations.length - 1, slot))
    const current = this.animations[normalizedSlot]
    if (current) {
      return current
    }

    const created = new SkeletonAnimator(normalizedSlot)
    this.animations[normalizedSlot] = created
    return created
  }
}
